File: lab4/GCD.java
//-----------------------------------------------------------
// GCD.java
// Shridhik John
// shjohn@ucsc.edu
// pa3
// A program that calculates the Greatest Common Denominator for 2 positive integers
//-----------------------------------------------------------

import java.util.Scanner;
class GCD{
   public static void main( String[] args ){
   int a, b, A, B, R;
    
        Scanner sc = new Scanner(System.in);
     
      // get two positive integers from the user

        System.out.print("Enter a positive integer: ");

        while(true)
        { // seemingly infinite loop 
            while( !sc.hasNextInt() )
            { // if its not a integer
            sc.next(); // discard the non-integer
            System.out.print("Please enter a positive integer: "); // ask again
            } // at this point we have a double in the input stream, unread
            a = sc.nextInt();
            if (a < 0)
                System.out.print("Please enter a positive integer: ");
            else 
                break;
        }
        
        System.out.print("Enter another positive integer: ");       
        while(true)
        { // seemingly infinite loop 
            while( !sc.hasNextInt() )
            { // if its not a double
            sc.next(); // discard the non-double
            System.out.print("Please enter a positive integer: "); // ask again
            } // at this point we have a double in the input stream, unread
                
            b = sc.nextInt();
            if (b < 0)
                System.out.print("Please enter a positive integer: ");
            else            
                break;       
        }

        A = Math.max(a,b); // finds the max value between the two inputs and stores max in A
        B = Math.min(a, b); // finds the min value between the two inputs and stores min in B

        
        while(B != 0) // When the remainder is not 0 the process will loop
        {
        
            R= A%B; // modulo to calculate theremainder between the  Min and Max 

            A=B; // assigns the devisor as the new dividend
            
            B=R; // assigns the remainder as the new divisor, and then checks the loop to see if the remainder is 0 yet
        
        }
    System.out.println("The GCD of "+ a +" and "+ b +" is " +A+"\n"); 
    }
}
                
          

 

File: lab8/GCD.c
/*
 * GCD.c
 * Shridhik John
 * shjohn@ucsc.edu
 * lab8
 * C program that prompts the user for two positive integers, then prints out the greatest common divisor of the two numbers.
*/

/* GCD.c */
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<ctype.h>

int isInteger(char s[]);
int max(int A,int B);
int min(int A,int B);

int main(){
	int i, n, a, A, B, R;
   	int x[2];
   	char str[100];
 
   	// get two positive integer from the user
         for(i=0; i<2; i++){
            if (i==0)
		   printf("Enter a positive integer: ");
	    else
                   printf("Enter another positive integer: ");

            n = scanf(" %s", str); // read one token from stdin
                           
            while( 1 ){ // seemingly infinite loop
               while( !isInteger(str) ){ // if its not a integer
                  printf("Please enter a positive integer: "); // ask again
                  n = scanf(" %s", str);             // get another string
               }
         
               // at this point str is a string that should be convertable to integer
         
               sscanf(str, "%d", &x[i]);  // convert str to integer and store in array x[]
               if( x[i]>0 ){               // if it's positive
                  break;                   // get out of the while loop
               }
               printf("Please enter a positive integer: "); // otherwise ask again
               n = scanf(" %s", str);                      // and get another
            }
            // break lands here
         }

	 A = max(x[0],x[1]);
	 B = min(x[0],x[1]);
        
	 while(B != 0){
           R= A%B;
           A=B;
           B=R;
         }

         // do the whole thing 3 times and end up here
         printf("The GCD of %d and %d is %d \n", x[0], x[1], A);
         return 0;
}




// Function to checks whether input is an integer
int isInteger(char s[]){
   int i, n;
   
   if( s==NULL ) 
      return 0;
   n = strlen(s);
   if( n==0 ) 
      return 0;
   if( s[0]!='-' && s[0]!='+' && !isdigit(s[0]) )
      return 0;
   for(i=1; i<n; i++){
      if( !isdigit(s[i]) ) 
         return 0;
   }
   return 1;
}


// functions that return max and min values 
int max(int A,int B){
  if (A > B)
	return A;
  else 
	return B;
}


int min(int A,int B){
 if (A < B) 
        return A;
 else 
        return B;
}

